# Data Analysis of free Apps

In this project I will be loking at dataset from google play store and apple store. This dataset will be used to gain some insights into what kind of free apps are customers into. Since free apps make profit through ads inside the apps, this small project would give some idea for developers where they could potentailly focus their resources to make profits.


In [1]:
from csv import reader
# opening the data and converting them to a list
opened_file = open(r'C:\Users\Lokesh\Desktop\Data Quest\Guided_Project\Project_1\googleplaystore.csv',encoding="utf8")
read_file = reader(opened_file)
google_store = list(read_file)
opened_file = open(r'C:\Users\Lokesh\Desktop\Data Quest\Guided_Project\Project_1\AppleStore.csv',encoding="utf8")
read_file = reader(opened_file)
apple_store = list(read_file)
#storing the data without headers
google_header = google_store[0]
google_data = google_store[1:]
apple_header = apple_store[0]
apple_data = apple_store[1:]

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

# Headers used in the dataset:
Description for the **play store** data can be found from this [link](https://www.kaggle.com/lava18/google-play-store-apps?select=googleplaystore.csv). While the **apple store** data can be found from this [link](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)

In [3]:
#google data = 10841 x 13
#apple data = 7197 x 16
#to find the header in each data
explore_data(google_store,0,1,True)
print('\n')
explore_data(apple_store,0,1,True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


Number of rows: 10842
Number of columns: 13


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


Number of rows: 7198
Number of columns: 16


# Cleaning the data
Deleting bad data which was highlighted under the discussions section of the dataset source

In [4]:
explore_data(google_data,10471,10473)
del(google_data[10472])

['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']




In [5]:
#finding duplicate entries
for app in google_data:
    name = app[0]
    if name == 'Facebook':
        print(app)

['Facebook', 'SOCIAL', '4.1', '78158306', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']
['Facebook', 'SOCIAL', '4.1', '78128208', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']


# Isolating apps that are repeated in data set based on maximum number of reviews

In [6]:
duplicate_apps = []
unique_apps = []

for app in google_data:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print('Total number of apps:',len(google_data))
print('Number of unique apps:',len(unique_apps))    
print('Number of duplicate apps:', len(duplicate_apps))
print('\n')

Total number of apps: 10840
Number of unique apps: 9659
Number of duplicate apps: 1181




In [7]:
reviews_max ={}
for row in google_data:
    name = row[0]
    n_reviews = float(row[3])
    if name in reviews_max  and (reviews_max[name] < n_reviews):
        reviews_max[name] = n_reviews
    if name not in reviews_max:
        reviews_max[name] = n_reviews

print(len(reviews_max))

9659


# Cleaning the list
Two empty lists are created:
* android_clean store the cleaned list which not only doesnt contain duplicate entries but also apps with hishest number of reviews.
* already_Added is created to account if duplicate entries contain same number of highest reviews but duplication is due to some other reason.

In [8]:
#empty lists to store cleaned data and app names
android_clean= [] #clean list
already_Added= [] # if data had multiple entries with same no of reviews taken at diff times
for row in google_data:
    name = row[0]
    n_reviews = float(row[3])
    
    if (n_reviews == reviews_max[name]) and (name not in already_Added):
        android_clean.append(row)
        already_Added.append(name)
        
print(len(android_clean))

9659


# Function to check non-english characters

In [9]:
def check(name):
    warn_index = 0
    for character in name:
        if ord(character) >127:
            warn_index +=1
    if warn_index>2:
        return False
    else:
        return True

# Sorting out non-english apps

In [10]:
android_english =[]
apple_english=[]

# Play store Data 
for row in android_clean:
    name = row[0]
    if check(name):
        android_english.append(row)
        
#App store data
for row in apple_data:
    name = row[0]
    if check(name):
        apple_english.append(row)

print(len(android_english))
print(len(apple_english))

9597
7197


# Isolating free apps

In [11]:
# android price is with index =7, apple index = 4
android_free = []
apple_free = []
for row in android_english:
    price = row[7]
    if price == '0' or price == '0.0':
        android_free.append(row)
        
for row in apple_english:
    price = row[4]
    if price == '0' or price == '0.0':
        apple_free.append(row)
        
print(len(android_free))
print(len(apple_free))

8848
4056


 Our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our app.
 To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

# Frequency of Data set :

In [12]:
# function that gives out frequencey table froa given dataset
print(apple_header)
#index =11 'prime genre'
print('\n')
print(google_header)
#index =1 -'Category', index = 9 - 'Genres'

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [13]:
#input dataset-list of lists, index-integer
#outputs freq table
def freq_table(dataset,index):
    table = {}
    #total is used to find percentage in next loop
    total = 0
    for row in dataset:
        value = row[index]
        total +=1
        if value in table:
            table[value] +=1
        else:
            table[value] =1        
    # percetange = table[value]/total *100
    #dictionary for storing percentage
    table_percent = {}
    for key in table:
        percent = ( table[key] / total ) * 100
        table_percent[key] = percent
        
    return table_percent



#funtion for sorting the frequenct table
def display_table(dataset, index):
    table = freq_table(dataset, index) #dictionary is returned
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key) #stored as tuple so that sorted() can be used
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
        


In [14]:
print('Apple Store data based on \'prime_genre\' column')
display_table(apple_free,11)

Apple Store data based on 'prime_genre' column
Games : 55.64595660749507
Entertainment : 8.234714003944774
Photo & Video : 4.117357001972387
Social Networking : 3.5256410256410255
Education : 3.2544378698224854
Shopping : 2.983234714003945
Utilities : 2.687376725838264
Lifestyle : 2.3175542406311638
Finance : 2.0710059171597637
Sports : 1.947731755424063
Health & Fitness : 1.8737672583826428
Music : 1.6518737672583828
Book : 1.6272189349112427
Productivity : 1.5285996055226825
News : 1.4299802761341223
Travel : 1.3806706114398422
Food & Drink : 1.0601577909270217
Weather : 0.7642998027613412
Reference : 0.4930966469428008
Navigation : 0.4930966469428008
Business : 0.4930966469428008
Catalogs : 0.22189349112426035
Medical : 0.19723865877712032


# Inference for Apple Store data:

We can see that among the free English apps, more than a half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.


In [15]:
print('Android data based on \'Category\' column')
display_table(android_free,1)
print('\n')
print('Android data based on \'Genres\' column')
display_table(android_free,9)

Android data based on 'Category' column
FAMILY : 18.942133815551536
GAME : 9.697106690777577
TOOLS : 8.453887884267631
BUSINESS : 4.599909584086799
PRODUCTIVITY : 3.899186256781193
LIFESTYLE : 3.887884267631103
FINANCE : 3.7070524412296564
MEDICAL : 3.5375226039783
SPORTS : 3.390596745027125
PERSONALIZATION : 3.322784810126582
COMMUNICATION : 3.2323688969258586
HEALTH_AND_FITNESS : 3.0854430379746836
PHOTOGRAPHY : 2.949819168173599
NEWS_AND_MAGAZINES : 2.802893309222423
SOCIAL : 2.667269439421338
TRAVEL_AND_LOCAL : 2.3395117540687163
SHOPPING : 2.2490958408679926
BOOKS_AND_REFERENCE : 2.1360759493670884
DATING : 1.8648282097649187
VIDEO_PLAYERS : 1.7970162748643763
MAPS_AND_NAVIGATION : 1.3901446654611211
FOOD_AND_DRINK : 1.2432188065099457
EDUCATION : 1.164104882459313
ENTERTAINMENT : 0.9606690777576853
LIBRARIES_AND_DEMO : 0.9380650994575045
AUTO_AND_VEHICLES : 0.9267631103074141
HOUSE_AND_HOME : 0.8024412296564195
WEATHER : 0.7911392405063291
EVENTS : 0.7120253164556962
PARENTING : 

# Average rating in Play Store

In [16]:
#'prime_genre'
genres_apple = freq_table(apple_free,-5)

for genre in genres_apple:
    total = 0 #to store sum of no. of user rating
    len_genre = 0 #to store no. of apps
    
    for row in apple_free:
        genre_app = row[-5]
        if genre_app == genre:
            rating = float(row[5])
            total += rating
            len_genre +=1
            
    avg_rating = total / len_genre
    print(genre,': ', avg_rating)

Social Networking :  53078.195804195806
Photo & Video :  27249.892215568863
Games :  18924.68896765618
Music :  56482.02985074627
Reference :  67447.9
Health & Fitness :  19952.315789473683
Weather :  47220.93548387097
Utilities :  14010.100917431193
Travel :  20216.01785714286
Shopping :  18746.677685950413
News :  15892.724137931034
Navigation :  25972.05
Lifestyle :  8978.308510638299
Entertainment :  10822.961077844311
Food & Drink :  20179.093023255813
Sports :  20128.974683544304
Book :  8498.333333333334
Finance :  13522.261904761905
Education :  6266.333333333333
Productivity :  19053.887096774193
Business :  6367.8
Catalogs :  1779.5555555555557
Medical :  459.75


Based on the average rating of each genre; **reference apps**, **music apps** and **social networking apps** have the highest number of user reviews. this simply means , these are the categories where people wrote reviews for the apps. This will give an insight for future development of the app , fixing bugs , etc.,

# Number of installs in play store in percentage

In [17]:
freq_table(android_free, 1) # the Installs columns


{'ART_AND_DESIGN': 0.6442133815551537,
 'AUTO_AND_VEHICLES': 0.9267631103074141,
 'BEAUTY': 0.599005424954792,
 'BOOKS_AND_REFERENCE': 2.1360759493670884,
 'BUSINESS': 4.599909584086799,
 'COMICS': 0.6103074141048824,
 'COMMUNICATION': 3.2323688969258586,
 'DATING': 1.8648282097649187,
 'EDUCATION': 1.164104882459313,
 'ENTERTAINMENT': 0.9606690777576853,
 'EVENTS': 0.7120253164556962,
 'FINANCE': 3.7070524412296564,
 'FOOD_AND_DRINK': 1.2432188065099457,
 'HEALTH_AND_FITNESS': 3.0854430379746836,
 'HOUSE_AND_HOME': 0.8024412296564195,
 'LIBRARIES_AND_DEMO': 0.9380650994575045,
 'LIFESTYLE': 3.887884267631103,
 'GAME': 9.697106690777577,
 'FAMILY': 18.942133815551536,
 'MEDICAL': 3.5375226039783,
 'SOCIAL': 2.667269439421338,
 'SHOPPING': 2.2490958408679926,
 'PHOTOGRAPHY': 2.949819168173599,
 'SPORTS': 3.390596745027125,
 'TRAVEL_AND_LOCAL': 2.3395117540687163,
 'TOOLS': 8.453887884267631,
 'PERSONALIZATION': 3.322784810126582,
 'PRODUCTIVITY': 3.899186256781193,
 'PARENTING': 0.65551


Based on the percetage of installs for each genre; **Family** has **18.9%** of total installs. This is followed by **Games** with **9.69%** and **Tools** with **8.45%**.

# Focussing on category column of the play store dataset

In [18]:
# 'category' in android daatset
category_android  = freq_table(android_free,1)
#print(category_android)

for category in category_android:
    total = 0
    len_category = 0
    for row in android_free:
        category_app = row[1]
        if category_app == category:
            n_installs = row[5]
            n_installs = n_installs.replace('+','')
            n_installs = n_installs.replace(',','')
            n_installs = float(n_installs)
            total += n_installs
            len_category += 1
    avg_installs = total / len_category
    print(category,":",avg_installs)    
            
    

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8814199.78835979
BUSINESS : 1712290.1474201474
COMICS : 832613.8888888889
COMMUNICATION : 38590581.08741259
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1360598.042253521
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1446158.2238372094
GAME : 15544014.51048951
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3650602.276666667
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10830251.970588235
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5145550.285714285
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_MAGAZ

# Inference for Google Play Store data:
The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.
The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.



# Conclusion:
Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. 